library/ssl_tls.c

```diff
@@ -3470,6 +3511,30 @@ int mbedtls_ssl_get_max_out_record_payload(const mbedtls_ssl_context *ssl)
     }
 #endif
 
+#if defined(MBEDTLS_SSL_RECORD_SIZE_LIMIT)
+    const size_t record_size_limit = mbedtls_ssl_get_output_record_size_limit(ssl);
+
+    if (max_len > record_size_limit) {
+        max_len = record_size_limit;
+    }
+#endif
+
+    if (ssl->transform_out != NULL &&
+        ssl->transform_out->tls_version == MBEDTLS_SSL_VERSION_TLS1_3) {
+        /* RFC 8449, section 4:
```

Comment:
It's not immediately clear why this is not within the `MBEDTLS_SSL_RECORD_SIZE_LIMIT` block.  Some of the text in [Ronald's comment](https://github.com/Mbed-TLS/mbedtls/pull/7455#discussion_r1433904997) might help here